internal/webhooks/patch_validation_test.go

```diff
@@ -0,0 +1,1372 @@
+							Required: true,
+							Schema: clusterv1.VariableSchema{
+								OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "error if jsonPatch defines neither ValueFrom.Template nor ValueFrom.Variable",
+			clusterClass: clusterv1.ClusterClass{
+				Spec: clusterv1.ClusterClassSpec{
+					ControlPlane: clusterv1.ControlPlaneClass{
+						LocalObjectTemplate: clusterv1.LocalObjectTemplate{
+							Ref: &corev1.ObjectReference{
+								APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+								Kind:       "ControlPlaneTemplate",
+							},
+						},
+					},
+					Patches: []clusterv1.ClusterClassPatch{
+						{
+							Name: "patch1",
+							Definitions: []clusterv1.PatchDefinition{
+								{
+									Selector: clusterv1.PatchSelector{
+										APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+										Kind:       "ControlPlaneTemplate",
+										MatchResources: clusterv1.PatchSelectorMatch{
+											ControlPlane: pointer.Bool(true),
+										},
+									},
+									JSONPatches: []clusterv1.JSONPatch{
+										{
+											Op:   "add",
+											Path: "/spec/template/spec/",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "error if jsonPatch has both ValueFrom.Template and ValueFrom.Variable",
+			clusterClass: clusterv1.ClusterClass{
+				Spec: clusterv1.ClusterClassSpec{
+					ControlPlane: clusterv1.ControlPlaneClass{
+						LocalObjectTemplate: clusterv1.LocalObjectTemplate{
+							Ref: &corev1.ObjectReference{
+								APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+								Kind:       "ControlPlaneTemplate",
+							},
+						},
+					},
+					Patches: []clusterv1.ClusterClassPatch{
+						{
+							Name: "patch1",
+							Definitions: []clusterv1.PatchDefinition{
+								{
+									Selector: clusterv1.PatchSelector{
+										APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+										Kind:       "ControlPlaneTemplate",
+										MatchResources: clusterv1.PatchSelectorMatch{
+											ControlPlane: pointer.Bool(true),
+										},
+									},
+									JSONPatches: []clusterv1.JSONPatch{
+										{
+											Op:   "add",
+											Path: "/spec/template/spec/",
+											ValueFrom: &clusterv1.JSONPatchValue{
+												Variable: pointer.String("variableName"),
+												Template: pointer.String(`template {{ .variableB }}`),
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					Variables: []clusterv1.ClusterClassVariable{
+						{
+							Name:     "variableName",
+							Required: true,
+							Schema: clusterv1.VariableSchema{
+								OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "pass if jsonPatch uses a valid index for add i.e. 0",
```

Comment:
Can we move those test cases below the other cases which are testing the rest of `validateJSONPatchValues`. It's hard to get a feel for the coverage if they are mixed